pkgs/development/tools/literate-programming/noweb/default.nix

```diff
@@ -1,29 +1,74 @@
-
-stdenv.mkDerivation {
-  name = "noweb-2.11b";
-  src = fetchurl {
-    urls = [ "http://ftp.de.debian.org/debian/pool/main/n/noweb/noweb_2.11b.orig.tar.gz"
-             "ftp://www.eecs.harvard.edu/pub/nr/noweb.tgz"
-          ];
-    sha256 = "10hdd6mrk26kyh4bnng4ah5h1pnanhsrhqa7qwqy6dyv3rng44y9";
+{ stdenv, fetchFromGitHub, gawk, groff, icon-lang ? null }:
+
+let noweb = stdenv.mkDerivation rec {
+  name = "${pname}-${version}";
+  version = "2.12";
+  pname = "noweb";
+  tlType = "run";
+
+  src = fetchFromGitHub {
+    owner = "nrnrnr";
+    repo = "noweb";
+    rev = "v${builtins.replaceStrings ["."] ["_"] version}";
+    sha256 = "1160i2ghgzqvnb44kgwd6s3p4jnk9668rmc15jlcwl7pdf3xqm95";
   };
+
+  outputs = [ "out" "tex" ];
+
+  nativeBuildInputs = [ groff ] ++ stdenv.lib.optionals (!isNull icon-lang) [ icon-lang ];
+
   preBuild = ''
-    mkdir -p $out/lib/noweb
+    mkdir -p "$out/lib/noweb"
     cd src
-    makeFlags="BIN=$out/bin LIB=$out/lib/noweb MAN=$out/share/man TEXINPUTS=$out/share/texmf/tex/latex"
   '';
-  preInstall=''mkdir -p $out/share/texmf/tex/latex'';
-  postInstall= ''
-    substituteInPlace $out/bin/cpif --replace "PATH=/bin:/usr/bin" ""
-    for f in $out/bin/{noweb,nountangle,noroots,noroff,noindex} \
-             $out/lib/noweb/{toroff,btdefn,totex,noidx,unmarkup,toascii,tohtml,emptydefn}; do
-      substituteInPlace $f --replace "nawk" "${gawk}/bin/awk"
+
+  makeFlags = stdenv.lib.optionals (!isNull icon-lang) [
+    "LIBSRC=icon"
+    "ICONC=icont"
+  ];
+
+  installFlags = [
+    "BIN=$(out)/bin"
+    "ELISP=$(out)/share/emacs/site-lisp"
+    "LIB=$(out)/lib/noweb"
+    "MAN=$(out)/share/man"
+    "TEXINPUTS=$(tex)/tex/latex/noweb"
+  ];
+
+  preInstall = ''
+    mkdir -p "$tex/tex/latex/noweb"
+  '';
+
+  installTargets = "install-code install-tex install-elisp";
+
+  postInstall = ''
+    substituteInPlace "$out/bin/cpif" --replace "PATH=/bin:/usr/bin" ""
+
+    for f in $out/bin/no{index,roff,roots,untangle,web} \
+             $out/lib/noweb/to{ascii,html,roff,tex} \
+             $out/lib/noweb/{bt,empty}defn \
+             $out/lib/noweb/{noidx,unmarkup}; do
+        # NOTE: substituteInPlace breaks Icon binaries, so make sure the script
+        #       uses (n)awk before calling.
+        if grep -q nawk "$f"; then
+            substituteInPlace "$f" --replace "nawk" "${gawk}/bin/awk"
+        fi
     done
+
+    # HACK: This is ugly, but functional.
+    PATH=$out/bin:$PATH make -BC xdoc
+    make $installFlags install-man
+
+    ln -s "$tex" "$out/share/texmf"
   '';
+
   patches = [ ./no-FAQ.patch ];
 
-  meta = {
-    platforms = stdenv.lib.platforms.linux;
+  meta = with stdenv.lib; {
+    description = "A simple, extensible literate-programming tool";
+    homepage = https://www.cs.tufts.edu/~nr/noweb;
+    license = licenses.bsd2;
+    maintainers = with maintainers; [ yurrriq ];
+    platforms = with platforms; linux ++ darwin;
   };
-}
+}; in noweb // { passthru.pkgs = [ noweb.tex ]; }
```

Comment:
This is still wrong. The `passthru.pkgs` needs to go in the attrset passed to `stdenv.mkDerivation`.

The whole point of `passthru` is everything in it is added to the resulting derivation at the very end, basically just like `// { … }` does, except since it's one of the attributes given to the call to `mkDerivation` it means overrides will preserve it.

For example, using `//`:

```sh
> nix eval '(with import <nixpkgs> {}; stdenv.mkDerivation { name = "foo"; } // { a = 1; }).a'
1
> # same thing but with an override applied
> nix eval '((with import <nixpkgs> {}; stdenv.mkDerivation { name = "foo"; } // { a = 1; }).overrideAttrs (_: {})).a'
error: attribute 'a' missing, at (string):1:1
```

And now using `passthru`:

```sh
> nix eval '(with import <nixpkgs> {}; stdenv.mkDerivation { name = "foo"; passthru.a = 1; }).a'
1
> # same thing but with an override applied
> nix eval '((with import <nixpkgs> {}; stdenv.mkDerivation { name = "foo"; passthru.a = 1; }).overrideAttrs (_: {})).a'
1
```

The reason for this is the resulting derivation from `stdenv.mkDerivation` has a function `overrideAttrs` which holds onto the original attrset passed to `stdenv.mkDerivation`, and when invoked it passes this set to the argument, merges the resulting attrset with the original one, and re-invokes `stdenv.mkDerivation`. As a result, anything you added to the set after the fact is discarded, because it wasn't captured in the original attrset retained by the call to `overrideAttrs`. This is why `passthru` exists, so anything defined this way is captured in that attrset and preserved across overrides.

Reply:
Thanks for the detailed explanation. I'm not sure how to accomplish this then, given the recursive nature of the derivation. Any advice?

Reply:
`let` bindings are visible inside their own definitions, so you can write

```nix
let noweb = stdenv.mkDerivation rec {
  # …
  passthru.pkgs = [ noweb.tex ];
  # …
}; in noweb
```

Reply:
Oh, cool! I had no idea. Thanks.

Reply:
If you ever need a recursive attrset where a `let` binding is awkward you can also use `lib.fix`:

```nix
lib.fix (self: stdenv.mkDerivation rec {
  # …
  passthru.pkgs = [ self.tex ];
  # …
})
```

But in this case the `let` binding is probably easier to understand.

Reply:
Awesome! Thanks so much for all this feedback. I'll definitely add that pattern to my repertoire.

Reply:
> Awesome! Thanks so much for all this feedback. I'll definitely add that pattern to my repertoire.

Finally true :smile: